File: packages/core/ios/Errors/GenericError.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation

/// Shared error contract for JS-facing promise rejections.
///
/// Raw values must match the JS `ErrorType` union in `@ping-identity/rn-types`.
public enum ErrorType: String {
  case argumentError = "argument_error"
  case authError = "auth_error"
  case bindingError = "binding_error"
  case davinciError = "davinci_error"
  case fidoError = "fido_error"
  case exchangeError = "exchange_error"
  case internalError = "internal_error"
  case networkError = "network_error"
  case parseError = "parse_error"
  case stateError = "state_error"
  case unknownError = "unknown_error"
  case wellknownError = "wellknown_error"
}

/// Serializable error payload shared across native modules.
///
/// - `type` classifies the error for app-level branching.
/// - `error` is a stable, module-specific code (for example, BROWSER_OPEN_ERROR).
/// - `message`, `code`, and `status` provide extra diagnostics when available.
public struct GenericError {
  public let type: ErrorType
  public let error: String
  public let message: String?
  public let code: Any?
  public let status: Any?

  public init(
    type: ErrorType,
    error: String,
    message: String? = nil,
    code: Any? = nil,
    status: Any? = nil
  ) {
    self.type = type
    self.error = error
    self.message = message
    self.code = code
    self.status = status
  }

  public func asDictionary() -> NSDictionary {
    let payload = NSMutableDictionary()
    payload["type"] = type.rawValue
    payload["error"] = error
    if let message {
      payload["message"] = message
    }
    if let code {
      payload["code"] = code
    }
    if let status {
      payload["status"] = status
    }
    return payload
  }

  public func asNSError(domain: String = "com.pingidentity", code: Int = 0) -> NSError {
    NSError(domain: domain, code: code, userInfo: asDictionary() as? [String: Any] ?? [:])
  }
}

/// Reject a React Native promise with a shared error payload.
///
/// Use this helper to ensure errors always conform to the shared contract.
public func reject(
  _ error: GenericError,
  rejecter: (String, String, NSError?) -> Void,
  underlyingError: NSError? = nil
) {
  let message = error.message ?? "Unknown error"
  rejecter(error.error, message, underlyingError ?? error.asNSError())
}

File: packages/core/android/src/main/java/com/pingidentity/rncore/error/GenericError.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rncore.error

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.WritableMap

/**
 * Shared error contract for JS-facing promise rejections.
 *
 * Values must match the JS `ErrorType` union in `@ping-identity/rn-types`.
 */
enum class ErrorType(val rawValue: String) {
    ARGUMENT_ERROR("argument_error"),
    AUTH_ERROR("auth_error"),
    BINDING_ERROR("binding_error"),
    DAVINCI_ERROR("davinci_error"),
    FIDO_ERROR("fido_error"),
    EXCHANGE_ERROR("exchange_error"),
    INTERNAL_ERROR("internal_error"),
    NETWORK_ERROR("network_error"),
    PARSE_ERROR("parse_error"),
    STATE_ERROR("state_error"),
    UNKNOWN_ERROR("unknown_error"),
    WELLKNOWN_ERROR("wellknown_error")
}

/**
 * Serializable error payload shared across native modules.
 *
 * - `type` classifies the error for app-level branching.
 * - `error` is a stable, module-specific code (e.g. BROWSER_OPEN_ERROR).
 * - `message`, `code`, and `status` provide extra diagnostics when available.
 */
data class GenericError(
    val type: ErrorType,
    val error: String,
    val message: String? = null,
    val code: Any? = null,
    val status: Any? = null
) {
    /**
     * Convert to a JS-friendly map for promise rejections.
     */
    fun toWritableMap(): WritableMap {
        val map = Arguments.createMap()
        map.putString("type", type.rawValue)
        map.putString("error", error)
        message?.let { map.putString("message", it) }
        code?.let { putValue(map, "code", it) }
        status?.let { putValue(map, "status", it) }
        return map
    }

    /**
     * Write a scalar value into the map with a best-effort type conversion.
     */
    private fun putValue(map: WritableMap, key: String, value: Any) {
        when (value) {
            is String -> map.putString(key, value)
            is Number -> map.putDouble(key, value.toDouble())
            is Boolean -> map.putBoolean(key, value)
            else -> map.putString(key, value.toString())
        }
    }
}

/**
 * Reject a promise using the shared error contract.
 *
 * The JS layer can rely on receiving the `GenericError` payload in the rejection.
 */
fun Promise.reject(error: GenericError, throwable: Throwable? = null) {
    reject(error.error, error.message, throwable, error.toWritableMap())
}
